From 449cdeea5da8e10857da2c0530f4c35e5bad04db Mon Sep 17 00:00:00 2001
From: David Howells <dhowells@redhat.com>
Date: Tue, 25 Nov 2014 16:35:47 -0500
Subject: [fs] overlayfs: embed root into overlay_readdir_data

Message-id: <20141125163547.18115.53514.stgit@warthog.procyon.org.uk>
Patchwork-id: 100938
O-Subject: [PATCH 38/47] overlayfs: embed root into overlay_readdir_data
Bugzilla: 985875
RH-Acked-by: Steve Best <sbest@redhat.com>

From: Al Viro <viro@zeniv.linux.org.uk>

no sense having it a pointer - all instances have it pointing to
local variable in the same stack frame

Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Original-commit: 49be4fb9cc3431fc4ebc71c764db848483b2a16c
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/overlayfs/readdir.c b/fs/overlayfs/readdir.c
index 0033095..453c4bf 100644
--- a/fs/overlayfs/readdir.c
+++ b/fs/overlayfs/readdir.c
@@ -41,7 +41,7 @@ struct dir_context {
 struct ovl_readdir_data {
  struct dir_context ctx;
  bool is_merge;
- struct rb_root *root;
+ struct rb_root root;
  struct list_head *list;
  struct list_head *middle;
  int count;
@@ -106,7 +106,7 @@ static int ovl_cache_entry_add_rb(struct ovl_readdir_data *rdd,
       const char *name, int len, u64 ino,
       unsigned int d_type)
 {
- struct rb_node **newp = &rdd->root->rb_node;
+ struct rb_node **newp = &rdd->root.rb_node;
  struct rb_node *parent = NULL;
  struct ovl_cache_entry *p;
 
@@ -131,7 +131,7 @@ static int ovl_cache_entry_add_rb(struct ovl_readdir_data *rdd,
 
  list_add_tail(&p->l_node, rdd->list);
  rb_link_node(&p->node, parent, newp);
- rb_insert_color(&p->node, rdd->root);
+ rb_insert_color(&p->node, &rdd->root);
 
  return 0;
 }
@@ -142,7 +142,7 @@ static int ovl_fill_lower(struct ovl_readdir_data *rdd,
 {
  struct ovl_cache_entry *p;
 
- p = ovl_cache_entry_find(rdd->root, name, namelen);
+ p = ovl_cache_entry_find(&rdd->root, name, namelen);
  if (p) {
   list_move_tail(&p->l_node, rdd->middle);
  } else {
@@ -284,12 +284,11 @@ static inline int ovl_dir_read_merged(struct path *upperpath,
           struct list_head *list)
 {
  int err;
- struct rb_root root = RB_ROOT;
  struct list_head middle;
  struct ovl_readdir_data rdd = {
   .ctx.actor = ovl_fill_merge,
   .list = list,
-  .root = &root,
+  .root = RB_ROOT,
   .is_merge = false,
  };
 
-- 
1.7.1